Wait for carrier before announcing IPs via GARP/NA

thom311 opened the pull request:
This is a follow up to c241dcb4367c.

what happens is that after sriov-cni sets up the interface, initially the interface does not have carrier. Sent GARP/IPv6 NA at that point are lost.

Instead, wait a bit until the interface has carrier.

See-also: https://issues.redhat.com/browse/OCPBUGS-30549

wizhaoredhat commented:
LGTM
@SchSeba @zeeke PTAL

thom311 commented:
@wizhaoredhat  previous version had a bug. Should be fixed now:
```diff
diff --git a/pkg/utils/packet.go b/pkg/utils/packet.go
index 809b4ded399f..88a52bc78cc1 100644
--- a/pkg/utils/packet.go
+++ b/pkg/utils/packet.go
@@ -221,7 +221,12 @@ func WaitForCarrier(ifName string, waitTime time.Duration) bool {
               return false
          }
 
-         if linkObj.Attrs().RawFlags&unix.IFF_NO_CARRIER != 0 {
+         /* Wait for carrier, i.e. IFF_UP|IFF_RUNNING. Note that there is also
+          * IFF_LOWER_UP, but we follow iproute2 ([1]).
+          *
+          * [1] https://git.kernel.org/pub/scm/network/iproute2/iproute2.git/tree/ip/ipaddress.c?id=f9601b10c21145f76c3d46c163bac39515ed2061#n86
+          */
+         if linkObj.Attrs().RawFlags&(unix.IFF_UP|unix.IFF_RUNNING) == (unix.IFF_UP | unix.IFF_RUNNING) {
               return true
          }
     }
```

zeeke commented on pkg/utils/packet.go:
```diff
@@ -196,3 +198,38 @@ func AnnounceIPs(ifName string, ipConfigs []*current.IPConfig) error {
+		 *
+		 * [1] https://git.kernel.org/pub/scm/network/iproute2/iproute2.git/tree/ip/ipaddress.c?id=f9601b10c21145f76c3d46c163bac39515ed2061#n86
+		 */
+		if linkObj.Attrs().RawFlags&(unix.IFF_UP|unix.IFF_RUNNING) == (unix.IFF_UP | unix.IFF_RUNNING) {
```
This will work if both IFF_UP and IFF_RUNNING flags are set. `IFF_RUNNING` seems to have a compatibility meaning only:

https://www.xml.com/ldd/chapter/book/ch14.html#t3
> IFF_RUNNING
This flag indicates that the interface is up and running. It is mostly present for BSD compatibility; the kernel makes little use of it. Most network drivers need not worry about IFF_RUNNING.

Is this the intended behavior?

thom311 replied:
I think checking for IFF_UP|IFF_RUNNING is correct.

Maybe an alternative would be to check for IFF_LOWER_UP, but according to https://www.kernel.org/doc/html/v5.9/networking/operstates.html#querying-from-userspace , I think IFF_RUNNING (and IFF_UP) is more appropriate. That is also what iproute2 does (as linked).

thom311 replied:
see for example: https://github.com/systemd/systemd/blob/77da8c4e9b167b1defe13896437389d8c219a926/src/shared/netif-util.c#L29

zeeke replied:
Agree, thanks for adding details here

zeeke reviewed (commented):
Please, consider adding a unit test for `WaitForCarrier` (see https://github.com/k8snetworkplumbingwg/sriov-cni/commit/f168bc2c4c41468425a6b296942a6d58cb48ee41)

thom311 commented:
> Please, consider adding a unit test for `WaitForCarrier` (see [f168bc2](https://github.com/k8snetworkplumbingwg/sriov-cni/commit/f168bc2c4c41468425a6b296942a6d58cb48ee41))

I took your patch and integrated it in the branch. How about this?

(I don't think the test is yet correct. WIP).

thom311 commented:
> (I don't think the test is yet correct. WIP).

Unit test should work now... IMO it's ready.

SchSeba commented:
Hi @thom311 the CI is still not happy can you please take a look?

thom311 commented:
> Hi @thom311 the CI is still not happy can you please take a look?

It says "race detected during execution of test".

Is this due to the assignment of `fakeLink.RawFlags` while another go routine reads it?

wizhaoredhat commented:
@adrianchiris Could you also help PTAL?

SchSeba commented:
Hi @thom311 yes that is the reason please take a look

```
WARNING: DATA RACE
Write at 0x00c000135bc8 by goroutine 49:
  github.com/k8snetworkplumbingwg/sriov-cni/pkg/utils.glob..func1.1.1()
      /home/runner/work/sriov-cni/sriov-cni/pkg/utils/packet_test.go:37 +0x4e4
  github.com/onsi/ginkgo/v2/internal.extractBodyFunction.func3()
      /root/go/pkg/mod/github.com/onsi/ginkgo/v2@v2.9.2/internal/node.go:463 +0x2e
  github.com/onsi/ginkgo/v2/internal.(*Suite).runNode.func3()
      /root/go/pkg/mod/github.com/onsi/ginkgo/v2@v2.9.2/internal/suite.go:863 +0x14b

Previous read at 0x00c000135bc8 by goroutine 50:
  github.com/k8snetworkplumbingwg/sriov-cni/pkg/utils.WaitForCarrier()
      /home/runner/work/sriov-cni/sriov-cni/pkg/utils/packet.go:225 +0xf0
  github.com/k8snetworkplumbingwg/sriov-cni/pkg/utils.glob..func1.1.1.2()
      /home/runner/work/sriov-cni/sriov-cni/pkg/utils/packet_test.go:32 +0x[44](https://github.com/k8snetworkplumbingwg/sriov-cni/actions/runs/9315930803/job/25643595721?pr=301#step:5:45)
```

@wizhaoredhat please first fix the unit-test so we can continue with the reviews on the PR.

adrianchiris commented on cmd/sriov/main.go:
```diff
@@ -209,6 +193,32 @@ func cmdAdd(args *skel.CmdArgs) error {
 		return fmt.Errorf("error saving the pci allocation for vf pci address %s: %v", netConf.DeviceID, err)
 	}
 
+	if doAnnounce {
```
why do we need to move this  logic here ? just to "give it more time" ?

thom311 replied:
To not do the non-criticial, potentially long blocking action in between.

The commit message attempts to and should explain that. Is that somehow unclear.

adrianchiris replied:
it is clear.

adrianchiris commented on pkg/utils/packet_test.go:
```diff
@@ -0,0 +1,42 @@
+	"time"
+
+	mocks_utils "github.com/k8snetworkplumbingwg/sriov-cni/pkg/utils/mocks"
+	"github.com/onsi/ginkgo/v2"
```
nit: maybe dot import ginkgo and gomega ?

thom311 replied:
I personally wouldn't. I also see that there are no dot imports anywhere else. But I don't mind, if that's the preferred style... (?).

adrianchiris replied:
it is preferred.

[1]https://github.com/k8snetworkplumbingwg/sriov-cni/blob/aacf339ed0c485c8edb22024e091a2720266c4cb/pkg/sriov/sriov_test.go#L13
[2] https://github.com/k8snetworkplumbingwg/sriov-cni/blob/aacf339ed0c485c8edb22024e091a2720266c4cb/pkg/utils/utils_test.go#L11

thom311 replied:
ah, I grepped for `"import \.", and missed those uses... Will update.

thom311 commented:
> Hi @thom311 yes that is the reason please take a look

Race in unit test should be fixed now. How is that.

adrianchiris commented:
@thom311 could you sort the PR ? i see it now got 167 commits (openshift fork?)

thom311 commented:
> @thom311 could you sort the PR commits? i see it now got 167 commits (openshift fork?)

sorry, that was a wrong rebase.

Fixed, and addressed other comments too.

zeeke reviewed (approved):
LGTM
